tiatoolbox/tools/registration/wsi_registration.py

```diff
@@ -0,0 +1,187 @@
+from numbers import Number
+from typing import Tuple, Union
+
+import cv2
+import numpy as np
+import scipy.ndimage as ndi
+import torchvision
+from skimage import exposure, filters, morphology
+from skimage.util import img_as_float
+from torchvision.models._utils import IntermediateLayerGetter
+
+Resolution = Union[Number, Tuple[Number, Number], np.ndarray]
+
+
+def preprocess(fixed_img, moving_img):
+    """This function performs normalization to unify the appearance of fixed
+     and moving images
+
+    Args:
+            fixed_img (:class:`numpy.ndarray`):
+                    A grayscale fixed image
+            moving_img (:class:`numpy.ndarray`):
+                    A grayscale moving image
+
+    Returns:
+            :class:`numpy.ndarray`:
+                    A normalized grayscale fixed image
+            :class:`numpy.ndarray`:
+                    A normalized grayscale moving image
+
+    """
+    if len(fixed_img.shape) != 2 or len(moving_img.shape) != 2:
+        raise ValueError(f'{"The input images should be grayscale images."}')
+
+    moving_entropy, fixed_entropy = filters.rank.entropy(
+        moving_img, morphology.disk(3)
+    ), filters.rank.entropy(fixed_img, morphology.disk(3))
+    if np.mean(fixed_entropy) > np.mean(moving_entropy):
+        moving_img = exposure.match_histograms(moving_img, fixed_img)
+    else:
+        fixed_img = exposure.match_histograms(fixed_img, moving_img)
+
+    return fixed_img, moving_img
+
+
+def dice(mask_1, mask_2):
+    """This function computes dice coefficent between two masks."""
+    if mask_1.shape != mask_2.shape:
+        x_size = max(mask_1.shape[1], mask_2.shape[1])
+        y_size = max(mask_1.shape[0], mask_2.shape[0])
+
+        right_y, right_x = y_size - mask_1.shape[0], x_size - mask_1.shape[1]
+        mask_1 = np.pad(mask_1, [(0, right_y), (0, right_x)], mode="constant")
+
+        right_y, right_x = y_size - mask_2.shape[0], x_size - mask_2.shape[1]
+        mask_2 = np.pad(mask_2, [(0, right_y), (0, right_x)], mode="constant")
+
+    mask_1 = mask_1.astype(np.bool)
+    mask_2 = mask_2.astype(np.bool)
+    return 2 * np.logical_and(mask_1, mask_2).sum() / (mask_1.sum() + mask_2.sum())
```

Comment:
@shaneahmed I think we may have discussed this before, but it didn't happen for one reason or another. I think we should set up a metrics module with functions for DICE and panoptic quality etc. I think it makes more sense to have it there than kept together with the WSI registration code.

Reply:
The dice function is now added to the tools.matrics.py file.